src/python/pants/backend/scala/util_rules/versions.py

```diff
@@ -3,14 +3,74 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
+from enum import Enum
 
+from pants.base.deprecated import warn_or_error
 from pants.engine.rules import collect_rules, rule
 from pants.jvm.resolve.common import Coordinate
+from pants.util.strutil import softwrap
+
+
+class InvalidScalaVersion(ValueError):
+    def __init__(self, scala_version: str) -> None:
+        super().__init__(
+            softwrap(
+                f"""Value '{scala_version}' is not a valid Scala version.
+            It should be formed of [major].[minor].[patch]"""
+            )
+        )
+
+
+class ScalaCrossVersionMode(Enum):
+    PARTIAL = "partial"
+    BINARY = "binary"
+    FULL = "full"
+
+    @classmethod
+    def from_str(cls, value: str) -> ScalaCrossVersionMode:
+        if value == ScalaCrossVersionMode.PARTIAL.value:
+            warn_or_error(
+                "2.21.0",
+                f"Scala cross version: {value}",
+                "Use value `binary` instead",
```

Comment:
When do you expect this deprecation to hit? Is it user facing, or only plugin-author facing?

Reply:
It could be either way but the deprecation warning is mostly motivated to user facing, as users can define a _cross-version_ mode using the `crossversion` field in the `scala_artifact` target. The motivation is down to the fact that in Scala-world the `partial` cross-version is, most of the time, dubbed as `binary`.

Reply:
Okay, thanks! Is there a more direct place you can put the deprecation? Such as in the relevant Field's constructor so that you can specify the `self.address`? This deprecation message will be hard for end-users to track down where the offending code is coming from.

Reply:
ok, I'll put in the target field